Update intro to Airflow guide (#179)
astronomer/airflow-guides: guides/intro-to-airflow.md

## History

> In 2015, Airbnb experienced a problem. They were growing like crazy and had a massive amount of data that was only getting larger. To achieve the vision of becoming a fully data-driven organization, they had to grow their workforce of data engineers, data scientists, and analysts - all of whom had to regularly work to automate processes by writing scheduled batch jobs. To satisfy the need for a robust scheduling tool, [Data Engineer Maxime Beauchemin](https://soundcloud.com/the-airflow-podcast/the-origins-of-airflow) created and open-sourced Airflow with the idea that it would allow them to quickly author, iterate on, and monitor their batch data pipelines.

Reviewer: ```suggestion
In 2015, Airbnb experienced a problem. They were growing like crazy and had a massive amount of data that was only getting larger. To achieve the vision of becoming a fully data-driven organization, they had to grow their workforce of data engineers, data scientists, and analysts — all of whom had to regularly automate processes by writing scheduled batch jobs. To satisfy the need for a robust scheduling tool,  [Maxime Beauchemin](https://soundcloud.com/the-airflow-podcast/the-origins-of-airflow) created and open-sourced Airflow with the idea that it would allow them to quickly author, iterate on, and monitor their batch data pipelines.
```

> Since Maxime's first commit way back then, Airflow has come a long way. The project joined the official Apache Foundation Incubator in April of 2016, where it lived and grew until it graduated as a top-level project on January 8th, 2019. Almost two years later, as of December 2020, Airflow has over 1,400 contributors, 11,230 commits, and 19,800 stars on Github. On December 17th 2020, [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) was released, bringing with it major upgrades and powerful new functionality. Airflow is used by thousands of Data Engineering teams around the world and is only getting more widely adopted as the community grows stronger. 

Reviewer: ```suggestion
Since Maxime's first commit way back then, Airflow has come a long way. The project joined the official Apache Foundation Incubator in April of 2016, where it lived and grew until it graduated as a top-level project on January 8th, 2019. Almost two years later, as of December 2020, Airflow has over 1,400 contributors, 11,230 commits, and 19,800 stars on Github. On December 17th 2020, [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) was released, bringing with it major upgrades and powerful new features. Airflow is used by thousands of Data Engineering teams around the world and continues to be adopted as the community grows stronger. 
```

## Overview

> [Apache Airflow](https://airflow.apache.org/index.html) is a platform for programmatically authoring, scheduling, and monitoring workflows. It is completely open-source and is especially useful in architecting and orchestrating complex data pipelines. Airflow was originally created to solve the issues that come with long-running cron tasks that execute hefty scripts, but has grown beyond that to be arguably the most flexible and powerful open-source data pipeline platform out there.

Reviewer: ```suggestion
[Apache Airflow](https://airflow.apache.org/index.html) is a platform for programmatically authoring, scheduling, and monitoring workflows. It is completely open source and is especially useful in architecting and orchestrating complex data pipelines. Airflow was originally created to solve the issues that come with long-running cron tasks and hefty scripts, but it's since grown to become one of the most powerful open source data pipeline platforms out there.
```

Reviewer: Good vibes only... no arguing

Airflow has a couple of key benefits, namely:

> - **It's dynamic:** everything is written in Python, so anything you can do in Python you can do in Airflow

Reviewer: ```suggestion
- **It's dynamic:** Anything you can do in Python, you can do in Airflow.
```

> - **It's extensible:** Airlfow has many plugins available out of the box that allow you to interact with most common external systems. You can also always create your own as needed.

Reviewer: ```suggestion
- **It's extensible:** Airflow has readily available plugins for interacting with most common external systems. You can also create your own plugins as needed.
```

Reviewer: It'd be nice to have information on how one might create plugins, or why it's easier to create plugins with Airflow than other platforms.

> - **It's scalable:** teams use Airflow's broad functionality to run thousands of tasks per day.

Reviewer: ```suggestion
- **It's scalable:** Teams use Airflow to run thousands of different tasks per day.
```

> With Airflow, workflows are architected and expressed as Directed Acyclic Graphs (DAGs), with each node of the DAG defined as a specific task. It is designed with the belief that all data pipelines are best expressed as code, and as such is a code-first platform that allows you to iterate on your workflows quickly and efficiently. As a result of its code-first design philosophy, Airflow allows for a degree of customizibility and extensibility that other pipeline tools do not support.

Reviewer: ```suggestion
With Airflow, workflows are architected and expressed as Directed Acyclic Graphs (DAGs), with each node of the DAG representing a specific task. Airflow is designed with the belief that all data pipelines are best expressed as code, and as such is a code-first platform where you can quickly iterate on workflows. This code-first design philosophy provides a degree of extensibility that other pipeline tools can't match.
```

## Use Cases

Airflow can be used for virtually any batch data pipelines, and there are a ton of [documented use cases](https://github.com/jghoman/awesome-apache-airflow#best-practices-lessons-learned-and-cool-use-cases) in the community. Because of its extensibility, Airflow is particularly powerful for orchestrating jobs with complex dependencies in multiple external systems.

> For example, the diagram below shows a complex use case that can easily be accomplished with Airflow. It highlights that by writing pipelines in code and leveraging the many available plugins, Airflow can easily integrate with any number of different, dependent systems and still provide a one-stop-shop for orchestration and monitoring.

Reviewer: ```suggestion
For example, the diagram below shows a complex use case that can easily be accomplished with Airflow. By writing pipelines in code and using Airflow's many available plugins, you can integrate with any number of different, dependent systems with just a single platform for orchestration and monitoring.
```

### DAG

> A Directed Acyclic Graph, or DAG, is a data pipeline defined in Python code. Each DAG represents a collection of all the tasks you want to run and is organized to show relationships between tasks directly in the Airflow UI. They are defined this way for the following reasons:

Reviewer: ```suggestion
A Directed Acyclic Graph, or DAG, is a data pipeline defined in Python code. Each DAG represents a collection of tasks you want to run and is organized to show relationships between tasks in Airflow's UI. When breaking down the properties of DAGs, their usefulness becomes clear:
```

> 1. Directed: If multiple tasks with dependencies exist, each must have at least one defined upstream or downstream task.

Reviewer: ```suggestion
* Directed: If multiple tasks with dependencies exist, each must have at least one defined upstream or downstream task.
```

> 2. Acyclic: Tasks are not allowed to create data that goes on to self-reference. This is to avoid creating infinite loops.

Reviewer: ```suggestion
* Acyclic: Tasks are not allowed to create data that goes on to self-reference. This is to avoid creating infinite loops.
```

> 3. Graph: All tasks are laid out in a clear structure with processes occurring at clear points with set relationships to other tasks.

Reviewer: ```suggestion
* Graph: All tasks are laid out in a clear structure with processes occurring at clear points with set relationships to other tasks.
```

![Example DAGs](https://assets2.astronomer.io/main/guides/intro-to-airflow/operator.png)

There are three main categories of operators:
 - **Action Operators** execute a function, like the [PythonOperator](https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/python.html) or [BashOperator](https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/bash.html)
 - **Transfer Operators** move data from a source to a destination, like the [S3ToRedshiftOperator](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/operators/s3_to_redshift.html)
 - **Sensor Operators** wait for something to happen, like the [ExternalTaskSensor](https://airflow.readthedocs.io/en/1.9.0/_modules/airflow/operators/sensors.html)

Operators are defined individually, but they can pass information to other operators using [XComs](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html?highlight=hook#xcoms).

> Taking DAGs, operators, and tasks together gives something conceptually like this:

Reviewer: ```suggestion
At a high level, the combined system of DAGs, operators, and tasks looks like this:
```

![Combined Concepts](https://assets2.astronomer.io/main/guides/intro-to-airflow/combined_concepts.png)

### Hooks

[Hooks](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html?highlight=hook#hooks) are Airflow's way of interfacing with third-party systems. They allow you to connect to external APIs and databases like Hive, S3, GCS, MySQL, Postgres, etc. They act as building blocks for operators. Secure information such as authentication credentials are kept out of hooks - that information is stored via Airflow connections in the encrypted metadata db that lives under your Airflow instance.

### Plugins

> Airflow plugins represent a combination of Hooks and Operators that allow you to accomplish a certain task, like [transfer data from Salesforce to Redshift](http://astronomer.io/guides/salesforce-to-redshift). Check out our [open-source library of Airflow plugins](https://github.com/airflow-plugins) if you'd like to check if a plugin you need has already been created by the community.

Reviewer: ```suggestion
Airflow plugins represent a combination of Hooks and Operators that can be used to accomplish a certain task, such as [transferring data from Salesforce to Redshift](http://astronomer.io/guides/salesforce-to-redshift). Check out our [open-source library of Airflow plugins](https://github.com/airflow-plugins) if you'd like to check if a plugin you need has already been created by the community.
```
Not sure how you feel about this, but I'm not a fan when a software "allows me" to do something. If this guide is meant to prove how Airflow can empower the user, allowing or letting a user do something — compared to to the user just doing it without the software's permission — feels overly prohibitive.

Author: Interesting, I've never thought about it that way, but I see your point. I'm totally fine with the update.
